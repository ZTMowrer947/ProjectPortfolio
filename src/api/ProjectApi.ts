// Imports
import { projects as projectData } from '@/../data.json';
import Project from '@/models/Project';

// Project data
const projects: Project[] = projectData.map((project) => ({
  id: project.id,
  name: project.project_name,
  description: project.description,
  technologies: project.technologies,
  liveLink: project.live_link ?? null,
  githubLink: project.github_link,
  landingImageUrl: project.image_urls.landing,
  galleryImageUrls: project.image_urls.gallery,
}));

// Service
class ProjectApi {
  public static async getList(): Promise<Project[]> {
    return Promise.resolve(projects);
  }

  public static async get(id: number): Promise<Project | undefined> {
    return Promise.resolve(projects.find((project) => project.id === id));
  }
}

// Exports
export default ProjectApi;
