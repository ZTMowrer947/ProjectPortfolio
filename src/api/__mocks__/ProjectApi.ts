// Imports
import { image, internet, lorem, random } from 'faker';
import { defer, Observable, of } from 'rxjs';

import Project from '@/models/Project';

// Mock data
const projects = Array.from(
  { length: 10 },
  (): Project => ({
    id: random.uuid(),
    name: random.words(3),
    description: lorem.paragraph(),
    technologiesUsed: Array.from({ length: 3 }, () => random.word()),
    liveLink: internet.url(),
    sourceLink: internet.url(),
    landingImage: image.imageUrl(),
    galleryImages: Array.from({ length: 3 }, () => image.imageUrl()),
  })
);

// Mock Service
class ProjectMockApi {
  public static getList(): Observable<Project[]> {
    return defer(() => of(projects));
  }

  public static get(id: string): Observable<Project | undefined> {
    return defer(() => of(projects.find((project) => project.id === id)));
  }
}

// Exports
export default ProjectMockApi;
