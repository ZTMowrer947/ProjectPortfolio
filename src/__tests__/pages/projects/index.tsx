// Imports
import { render } from '@testing-library/react';

import ProjectApi from '@/api/ProjectApi';
import Portfolio from '@/pages/projects';

// Jest mocks
jest.mock('@/api/ProjectApi');

// Test Suite
describe('<Portfolio>', () => {
  it('should render a list of project items', async () => {
    // Get mock project listing
    const projects = await ProjectApi.getList().toPromise();

    // Render component
    const { getByText, getAllByTestId } = render(
      <Portfolio projects={projects} />
    );

    // Get all project items
    const projectItems = getAllByTestId('project-item');

    // Expect there to be a project item for each project
    expect(projectItems).toHaveLength(projects.length);

    // For each project item,
    projectItems.forEach((projectItem, index) => {
      // Get corresponding project
      const project = projects[index];

      // Get project link, title, and landingImage
      const projectLink = projectItem.querySelector('a');
      const projectTitle = getByText(project.name);
      const projectLandingImage = projectItem.querySelector('img');

      /* eslint-disable @typescript-eslint/no-unsafe-call */

      // Expect all three to exist and be in the document
      expect(projectLink).toBeInTheDocument();
      expect(projectTitle).toBeInTheDocument();
      expect(projectLandingImage).toBeInTheDocument();

      // Expect link href to end with project ID
      expect(projectLink).toHaveAttribute(
        'href',
        expect.stringMatching(new RegExp(`${project.id}$`))
      );

      // Expect image source to match landing image URL for project
      expect(projectLandingImage).toHaveAttribute('src', project.landingImage);

      /* eslint-enable @typescript-eslint/no-unsafe-call */
    });
  });
});
