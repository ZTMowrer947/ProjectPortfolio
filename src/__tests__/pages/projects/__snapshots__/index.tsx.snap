// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio page should render a list of links to projects 1`] = `
Array [
  <article>
    <h1
      className="display-4"
    >
      Welcome!
    </h1>
    <p
      className="lead text-dark"
    >
      Below is the many projects demonstrating my programming prowess, primarily in Full-Stack JavaScript. Many of the below projects also have live demos, the links to which are provided in the detail page for such projects.
    </p>
  </article>,
  <hr />,
  <article>
    <ul
      className="list-unstyled row"
    >
      <li
        className="col-lg-4 col-md-6 col-12"
        data-testid="project-item"
      >
        <a
          aria-label="project-945bc390-1a6e-40c5-8c22-5a2011aa975c"
          className="text-decoration-none"
          href="/projects/945bc390-1a6e-40c5-8c22-5a2011aa975c"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <img
            alt="Landing image for \\"Project #1\\""
            className="img-fluid img-thumbnail"
            src="https://place-hold.it/640x480"
          />
          <span
            className="sc-bdfBwQ dHJqBU project-title mb-4 d-inline-block text-center text-uppercase w-100"
          >
            Project #1
          </span>
        </a>
      </li>
      <li
        className="col-lg-4 col-md-6 col-12"
        data-testid="project-item"
      >
        <a
          aria-label="project-98ab5f5f-7284-4101-88d9-856d4d584af0"
          className="text-decoration-none"
          href="/projects/98ab5f5f-7284-4101-88d9-856d4d584af0"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <img
            alt="Landing image for \\"Project #2\\""
            className="img-fluid img-thumbnail"
            src="https://place-hold.it/640x480"
          />
          <span
            className="sc-bdfBwQ dHJqBU project-title mb-4 d-inline-block text-center text-uppercase w-100"
          >
            Project #2
          </span>
        </a>
      </li>
      <li
        className="col-lg-4 col-md-6 col-12"
        data-testid="project-item"
      >
        <a
          aria-label="project-cdbcf90e-d02d-4735-9b91-4313bcb9e642"
          className="text-decoration-none"
          href="/projects/cdbcf90e-d02d-4735-9b91-4313bcb9e642"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <img
            alt="Landing image for \\"Project #3\\""
            className="img-fluid img-thumbnail"
            src="https://place-hold.it/640x480"
          />
          <span
            className="sc-bdfBwQ dHJqBU project-title mb-4 d-inline-block text-center text-uppercase w-100"
          >
            Project #3
          </span>
        </a>
      </li>
    </ul>
  </article>,
]
`;
